### PR TITLE
async inbox load

### PR DESCRIPTION
**Pull request by chrisnojima**
This enables the async inbox loading
We first load a server trusted list of the inbox, we then stream in unboxed inbox items and mark them validated. you can only load a validated conversation

### CODE REVIEW

File: shared/actions/chat.js [hunk 1]
```diff
@@ -42,11 +43,18 @@ import type {
...
   UpdateBadging,
+  UpdateInbox,
   UpdateLatestMessage,
   UpdateMetadata,
 } from '../constants/chat'
 
-const {conversationIDToKey, keyToConversationID, InboxStateRecord, makeSnippet, MetaDataRecord} = Constants
+const {conversationIDToKey, keyToConversationID, InboxStateRecord, MetaDataRecord, makeSnippet} = Constants
+
+const _selectedSelector = (state: TypedState) => state.chat.get('selectedConversation')
```
Comment (chrisnojima) [inline shared/actions/chat.js]: we use these selectors multiple times so pulling them up

File: shared/actions/chat.js [hunk 2]
```diff
@@ -101,45 +109,75 @@ function selectConversation (conversationIDKey: ConversationIDKey, fromUser: boo
   return {type: Constants.selectConversation, payload: {conversationIDKey, fromUser}}
 }
 
-function _inboxToConversations (inbox: GetInboxAndUnboxLocalRes, author: ?string, following: {[key: string]: boolean}): List<InboxState> {
-  return List((inbox.conversations || []).map(convo => {
-    if (!convo.info.id) {
-      return null
-    }
+function _inboxConversationToConversation (convo: ConversationLocal, author: ?string, following: {[key: string]: boolean}): ?InboxState {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: this loads the trusted inbox item

File: shared/actions/chat.js [hunk 3]
```diff
@@ -101,45 +109,75 @@ function selectConversation (conversationIDKey: ConversationIDKey, fromUser: boo
...
+    participants,
+    muted: false,
+    time: convo.readerInfo.mtime,
+    snippet,
+    unreadCount: convo.readerInfo.maxMsgid - convo.readerInfo.readMsgid,
+    validated: true,
+  })
+}
+
+function _inboxToConversations (inbox: GetInboxLocalRes, author: ?string, following: {[key: string]: boolean}): List<InboxState> {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: this loads the untrusted inbox item

File: shared/actions/chat.js [hunk 4]
```diff
@@ -342,23 +384,66 @@ function * _setupNewChatHandler (): SagaGenerator<any, any> {
...
-    query: {
-      status: Object.keys(CommonConversationStatus).filter(k => !['ignored', 'blocked'].includes(k)).map(k => CommonConversationStatus[k]),
+  const channelConfig = singleFixedChannelConfig([
+    'chat.1.chatUi.chatInboxUnverified',
+    'chat.1.chatUi.chatInboxConversation',
+    'chat.1.chatUi.chatInboxFailed',
+    'finished',
+  ])
+
+  const loadInboxChanMap: ChannelMap<any> = localGetInboxNonblockLocalRpcChannelMap(channelConfig, {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: Load the unverified inbox and deal with the stream of validations / errors. When the rpc is finished we clear out anything that didn't validate

File: shared/actions/chat.js [hunk 5]
```diff
@@ -608,7 +706,7 @@ function * _updateBadging (action: UpdateBadging): SagaGenerator<any, any> {
   const {conversationIDKey} = action.payload
   const conversationStateSelector = (state: TypedState) => state.chat.get('conversationStates', Map()).get(conversationIDKey)
   const conversationState = yield select(conversationStateSelector)
-  if (conversationState && conversationState.messages !== null) {
+  if (conversationState && conversationState.messages !== null && conversationState.messages.size > 0) {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: more sanity checking

File: shared/chat/conversation/list.desktop.js [hunk 6]
```diff
@@ -55,17 +55,7 @@ class ConversationList extends Component<void, Props, State> {
...
-        if (message.type === 'Timestamp') {
-          id = message.timestamp
-        } else {
-          // We want a stable key -- messages have an outboxID but no
-          // messageID, then later gain a messageID.  So if we prefer
-          // outboxIDs to messageIDs for the key, every row keeps its key.
-          id = message && (message.outboxID || message.messageID)
-        }
-      }
+      const id = message && message.key
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: we have an explicit key now

File: shared/chat/conversation/list.desktop.js [hunk 7]
```diff
@@ -210,6 +200,13 @@ class ConversationList extends Component<void, Props, State> {
   }
 
   render () {
+    if (!this.props.validated) {
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: Show a loading screen

File: shared/chat/conversations-list/index.desktop.js [hunk 8]
```diff
@@ -40,19 +40,22 @@ const Row = ({onSelectConversation, selectedConversation, onNewChat, nowOverride
...
   // $FlowIssue
   const avatarProps = participants.slice(0, 2).map((p, idx) => ({
     backgroundColor: globalColors.darkBlue4,
     username: p.username,
-    borderColor: rowBorderColor(idx, Math.min(2, participants.count()) - 1, !!conversation.get('unreadCount'), isSelected),
+    borderColor: rowBorderColor(idx, Math.min(2, participants.count()) - 1, hasUnread, isSelected),
   })).toArray()
   const snippet = conversation.get('snippet')
-  const subColor = isSelected ? globalColors.white : globalColors.blue3_40
+  const subColor = (isSelected || hasUnread) ? globalColors.white : globalColors.blue3_40
```
Comment (chrisnojima) [inline shared/chat/conversations-list/index.desktop.js]: There are effectively 3 states now. Selected, not selected, not selected + badged

File: shared/reducers/chat.js [hunk 9]
```diff
@@ -2,26 +2,24 @@
...
 
 const {StateRecord, ConversationStateRecord, makeSnippet} = Constants
 const initialState: State = new StateRecord()
 const initialConversation: ConversationState = new ConversationStateRecord()
 
-function dedupeMessages (seenMessages: Set<MessageID>, knownMessages: List<ServerMessage>, newMessages: Array<ServerMessage>): {seenMessages: Set<MessageID>, updatedMessages: List<ServerMessage>, unseenMessages: List<ServerMessage>} {
-  const newButSeenMessages = newMessages
-    .filter(m => m.messageID && seenMessages.has(m.messageID))
-    .reduce((acc, m) => acc.set(m.messageID, m), Map())
+function _dedupeMessages (seenMessages: Set<any>, messages: List<ServerMessage> = List(), prepend: List<ServerMessage> = List(), append: List<ServerMessage> = List()): {nextSeenMessages: Set<any>, nextMessages: List<ServerMessage>} {
```
Comment (chrisnojima) [inline shared/reducers/chat.js]: switch to using keys and always concat prepend/append

**Comment by chrisnojima** [comment]
@keybase/react-hackers

**Comment by keybase-ci-visdiff** [comment]
The commits 40d4abd605bc...da09295a5275 introduce visual changes on linux. <details><summary>:mag_right: 6 changed</summary>

 * Changed: **Chat_List-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_List-Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_List-Normal-da09295a5275.png)  
   Chat_List-Normal comparison
 * Changed: **Chat_List-Empty** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_List-Empty-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_List-Empty-da09295a5275.png)  
   Chat_List-Empty comparison
 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Input-Emoji_Open-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Input-Emoji_Open-da09295a5275.png)  
   Chat_Input-Emoji_Open comparison
 * Changed: **Chat_Conversations_List-Selected_Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Conversations_List-Selected_Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Conversations_List-Selected_Normal-da09295a5275.png)  
   Chat_Conversations_List-Selected_Normal comparison
 * Changed: **Chat_Conversations_List-Selected_Muted** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Conversations_List-Selected_Muted-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Conversations_List-Selected_Muted-da09295a5275.png)  
   Chat_Conversations_List-Selected_Muted comparison
 * Changed: **Chat_Conversations_List-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Conversations_List-Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481738783516-40d4abd605bc-da09295a5275-linux/Chat_Conversations_List-Normal-da09295a5275.png)  
   Chat_Conversations_List-Normal comparison
</summary>

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

File: shared/actions/chat.js [hunk 10]
```diff
@@ -342,23 +384,66 @@ function * _setupNewChatHandler (): SagaGenerator<any, any> {
...
+  const inbox: GetInboxLocalRes = chatInboxUnverified.params.inbox
+  const author = yield select(usernameSelector)
+  const following = yield select(followingSelector)
   const conversations: List<InboxState> = _inboxToConversations(inbox, author, following || {})
   yield put({type: Constants.loadedInbox, payload: {inbox: conversations}})
+  chatInboxUnverified.response.result()
+
+  // +1 for the finish call
+  const total = inbox.conversationsUnverified && inbox.conversationsUnverified.length + 1 || 0
+  for (let i = 0; i < total; ++i) {
```
Comment (cjb) [inline shared/actions/chat.js]: surprising there isn't an ES6 `for (let i of range(0, total)) { .. }` kind of thing!  Looks like the closest would be:
```js
for (let i of Array(total).keys()) { .. }
```

Comment (cjb) [inline shared/actions/chat.js]: Or, `_.range(total)`.

**Comment by chrisnojima** [comment]
Just to document this, the validation is the daemon async crypto checking the inbox. The initial load (the unverified rpc) is server trusted (just gets the names / timestamp). Then we get validations or fails for each inbox item. When the rpc is done we'll toss anything that doesn't validate

**Comment by keybase-ci-visdiff** [comment]
The commits 40d4abd605bc...a091a379eaf8 introduce visual changes on linux. <details><summary>:mag_right: 8 changed</summary>

 * Changed: **Chat_List-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_List-Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_List-Normal-a091a379eaf8.png)  
   Chat_List-Normal comparison
 * Changed: **Chat_List-Empty** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_List-Empty-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_List-Empty-a091a379eaf8.png)  
   Chat_List-Empty comparison
 * Changed: **Chat_Input-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Input-Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Input-Normal-a091a379eaf8.png)  
   Chat_Input-Normal comparison
 * Changed: **Chat_Input-Empty** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Input-Empty-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Input-Empty-a091a379eaf8.png)  
   Chat_Input-Empty comparison
 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Input-Emoji_Open-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Input-Emoji_Open-a091a379eaf8.png)  
   Chat_Input-Emoji_Open comparison
 * Changed: **Chat_Conversations_List-Selected_Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Selected_Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Selected_Normal-a091a379eaf8.png)  
   Chat_Conversations_List-Selected_Normal comparison
 * Changed: **Chat_Conversations_List-Selected_Muted** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Selected_Muted-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Selected_Muted-a091a379eaf8.png)  
   Chat_Conversations_List-Selected_Muted comparison
 * Changed: **Chat_Conversations_List-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Normal-40d4abd605bc.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Normal-a091a379eaf8.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481820347927-40d4abd605bc-a091a379eaf8-linux/Chat_Conversations_List-Normal.png)
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits ee6fc47de357...5b2081df08dd introduce visual changes on linux. <details><summary>:mag_right: 8 changed</summary>

 * Changed: **Menubar-Private_Async_Writing** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Menubar-Private_Async_Writing-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Menubar-Private_Async_Writing-5b2081df08dd.png)  
   Menubar-Private_Async_Writing comparison
 * Changed: **Chat_List-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_List-Normal-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_List-Normal-5b2081df08dd.png)  
   Chat_List-Normal comparison
 * Changed: **Chat_List-Empty** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_List-Empty-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_List-Empty-5b2081df08dd.png)  
   Chat_List-Empty comparison
 * Changed: **Chat_Input-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Input-Normal-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Input-Normal-5b2081df08dd.png)  
   Chat_Input-Normal comparison
 * Changed: **Chat_Input-Empty** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Input-Empty-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Input-Empty-5b2081df08dd.png)  
   Chat_Input-Empty comparison
 * Changed: **Chat_Conversations_List-Selected_Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Selected_Normal-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Selected_Normal-5b2081df08dd.png)  
   Chat_Conversations_List-Selected_Normal comparison
 * Changed: **Chat_Conversations_List-Selected_Muted** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Selected_Muted-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Selected_Muted-5b2081df08dd.png)  
   Chat_Conversations_List-Selected_Muted comparison
 * Changed: **Chat_Conversations_List-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Normal-ee6fc47de357.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Normal-5b2081df08dd.png) [(diff)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481824766793-ee6fc47de357-5b2081df08dd-linux/Chat_Conversations_List-Normal.png)
</summary>

**Pull request merged**